### PR TITLE
FEAT - Gantt Chart Rendering Type

### PR DESCRIPTION
**Pull request by osshad**
## Description
Created a Gantt Chart rendering type (`TimeSeriesType.GanttChart`) which shows the status of different events between a specified time range

image

```
res.Dataset.Add(new DiagnosticData()
   {
       Table = <datatable>,
       RenderingProperties = new TimeSeriesRendering(){
           Title = <title>,
           GraphType = TimeSeriesType.GanttChart, // required
           GraphOptions = new {
               timeGrain = <time_grain>, // required
               // taskBar object is optional and should only be included if we want to customize the task bar
               taskBar = new {
                   useCustomColor = true,
                   customColorColumnName = <column_name_for_customized_color>,
                   customTooltipColumnName = <column_name_for_customized_tooltip>,
                   customColorDescriptionColumnName = <column_name_for_description_of_the_customized_color>
               }
           },
           SeriesColumns = new string[] { <column_name_of_the_values_for_the_chart's_yAxis>" }, // required
           TimestampColumnName = <column_name_for_timestamp>, // This is not required if the table only has one column w/ type datetime
           EventStatusColumnName = <column_name_for_each_event_status> // required
```
Work Item: [#2632
](https://dev.azure.com/app-service-diagnostics-portal/app-service-diagnostics-portal/_workitems/edit/2632)

### This PR
 - [x] creates a Gantt Chart rendering type
 - [x] adds a new field `eventStatusColumnName` to the TimeSeriesRendering interface
 - [x] reverses the yAxis if the rendering type is Gantt Chart
 - [x] adds a new member `GanttChart` to the TimeSeriesType enum
 - [x] adds a new function which processes diagnostic data for Gantt Chart
 - [x] updates the `TimeSeriesGraph` and `HighCharts-Graph` components to render the gantt chart
 - [x] optionally renders markdown data above and/or below the gantt chart. This is only necessary for the instance allocation Event detector
 - [x] adds a new enum `GanttChartInnerMarkdownPosition` for specifying the position of the inner markdown (above or below the gantt chart)
 - [x] adds a new component `gantt-chart-taskbar-colors` for describing what the taskbar colors of the gantt chart mean (when custom colors are used)
 

### To Test
 - You'll need to run Diagnostics.RuntimeHost on [this branch](https://msazure.visualstudio.com/One/_git/AAPT-Antares-AppServiceDiagnostics?version=GBosshad%2Fgantt_chart)
 - In your terminal, run the following commands
       - `cd AngularApp` 
       - `npm run applens` 
 - Go to localhost:4200 and provide an app name e.g. buggybakery
 - Create a new detector, add the following code snippet to the editor and run it
 ```
[AppFilter(AppType = AppType.WebApp, PlatformType = PlatformType.Windows, StackType = StackType.All)]
[Definition(Id = "gantt_chart_rendering", Name = "gantt chart rendering", Author = "olu", Description = "")]
public async static Task<Response> Run(DataProviders dp, OperationContext<App> cxt, Response res)
{
    res.Dataset.Add(new DiagnosticData()
    {
        Table = createTableForTesting(),
        RenderingProperties = new TimeSeriesRendering(){
            Title = "Testing",
            GraphType = TimeSeriesType.GanttChart,
            GraphOptions = new {
                timeGrain = 10,
                taskBar = new {
                    useCustomColor = true,
                    customColorColumnName = "StatusColor",
                    customTooltipColumnName = "MachineName",
                    customColorDescriptionColumnName = "Status"
                }
            },
            SeriesColumns = new string[] { "MachineName" },
            TimestampColumnName = "PrecisionTime",
            EventStatusColumnName = "Status"
        }
    });

    return res;
}

private static DataTable createTableForTesting()
{
    DataTable customTable = new DataTable();
    customTable.Columns.Add("Id", typeof(int));
    customTable.Columns.Add("MachineName", typeof(string));
    customTable.Columns.Add("PrecisionTime", typeof(DateTime));
    customTable.Columns.Add("Status", typeof(string));
    customTable.Columns.Add("StatusColor", typeof(string));

    var colors = new Dictionary<string, string> {
        {"active", "#48FD48"},
        {"inactive", "#FD6D48"}
    };

    var machines = new Dictionary<string, List<string>> { 
        {"machine1", new List<string> { "active", "inactive", "active"} },
        {"machine2", new List<string> { "inactive", "inactive", "active"} },
        {"machine3", new List<string> { "active", "active", "active"} },
        {"machine4", new List<string> { "inactive", "inactive", "inactive"} },
    };
    
    DateTime utcDate = DateTime.UtcNow;
    int index = 1;
    foreach(KeyValuePair<string, List<string>> entry in machines)
    {
        var key = entry.Key;
        foreach(string value in entry.Value)
        {
            customTable.Rows.Add(index++, key, utcDate, value, colors[value]);
            utcDate = utcDate.AddMinutes(10);
        }
    }

    foreach(string value in machines["machine1"])
    {
        customTable.Rows.Add(index++, "machine1", utcDate, value, colors[value]);
        utcDate = utcDate.AddMinutes(10);
    }

    return customTable;
}
```

### Extra Notes
 - If the user plan to also render markdown data above and/or below the Gantt chart, the user can add the innerMarkdownData attribute to the `GraphOptions` object like this:
 ```
GraphOptions = new {
    timeGrain = Time_Grain,
    taskBar = new {
        useCustomColor = true,
        customColorColumnName = "StatusColor",
        customTooltipColumnName = "MachineName",
        customColorDescriptionColumnName = "Status"
    },
    innerMarkdownData = new [] {
        new { markdownData = operationEventHtml, Position = GanttChartInnerMarkdownPosition.AboveChart },
        new { markdownData = serverFarmAllocationHTML, Position = GanttChartInnerMarkdownPosition.BelowChart },
        new { markdownData = duplicateTenantInstanceNote, Position = GanttChartInnerMarkdownPosition.BelowChart }
     }
}
``` 

_**An example of a detector where this is applicable is the Instance Allocations Event detector**_

### CODE REVIEW

File: AngularApp/projects/applens/src/app/app.module.ts [hunk 1]
```diff
@@ -173,7 +173,8 @@ export const Routes = RouterModule.forRoot([
     FabPanelModule,
     FabDialogModule,
     FabButtonModule,
-    FabChoiceGroupModule
+    FabChoiceGroupModule,
+    HighchartsChartModule,
```
Comment (SharmaAjay19) [inline AngularApp/projects/applens/src/app/app.module.ts:177]: We should not require an explicit import of this module. It is already being used to render all our existing graphs. Try running once without this import.

Comment (SharmaAjay19) [reply AngularApp/projects/applens/src/app/app.module.ts:177]: The import is done in diagnostic-data module is my guess.

**Review by SharmaAjay19 — APPROVED** [review]

File: AngularApp/projects/applens/src/app/app.module.ts (hunk 1 above)
Comment (osshad) [reply AngularApp/projects/applens/src/app/app.module.ts:177]: That's true. We don't need it. I'll remove it now

**Review by ShekharGupta1988 — APPROVED** [review]

**Comment by ShekharGupta1988** [comment]
Very nice @osshad .. Work with Hanna and Sukhwant from our PM team to also update the docs for this new chart.

**Review by nmallick1 — APPROVED** [review]

**Review by rekhaswaminathan — APPROVED** [review]

File: AngularApp/projects/diagnostic-data/src/lib/components/time-series-graph/time-series-graph.component.ts [hunk 2]
```diff
@@ -325,8 +331,19 @@ export class TimeSeriesGraphComponent extends DataRenderBaseComponent implements
     }
 
     private _getDifferenceInMinutes(currentTime: string, endTime: string) {
-        const diffInMilliseconds = +new Date(currentTime) - +new Date(endTime);
-        return Math.round(((diffInMilliseconds % 86400000) % 3600000) / 60000);
+        const diff =
+          (this._getTimeStamp(currentTime) - this._getTimeStamp(endTime)) /
```
Comment (osshad) [inline AngularApp/projects/diagnostic-data/src/lib/components/time-series-graph/time-series-graph.component.ts:335]: These code changes were added to ensure that the differences in minutes were properly calculated.

File: AngularApp/projects/diagnostic-data/src/lib/components/time-series-graph/time-series-graph.component.ts [hunk 3]
```diff
@@ -114,26 +114,32 @@ export class TimeSeriesGraphComponent extends DataRenderBaseComponent implements
...
-        } else {
-            this.mappedGanttChartData[key] = [
-            {
-                ...data,
-                [this.timestampColumnName]: momentNs.utc(currentTime).toDate(),
-                EndTime: momentNs.utc(currentTime).add(timeGrain, 'm').toDate(),
-            },
-            ];
-        }
+                chartRowData.push({
```
Comment (osshad) [inline AngularApp/projects/diagnostic-data/src/lib/components/time-series-graph/time-series-graph.component.ts:117]: Here, I changed `toDate()` to `toISOString()` and formatted the code

**Comment by osshad** [comment]
> Very nice @osshad .. Work with Hanna and Sukhwant from our PM team to also update the docs for this new chart.

Thank you Shekhar. I'll reach out to Hanna and Sukhwant asap

**Pull request merged**